### PR TITLE
Fix http.server.active_requests metric with async requests

### PR DESCRIPTION
**Pull request by laurit**
Resolves https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/11625
With async requests `start` is called from the server instrumentation and `end` from servlet instrumentation. `http.server.active_requests` metric, that is recorded both in start and end, produces 2 metrics. The metric with server instrumentation scope has value `1` and the one with servlet instrumentation scope has value `-1` instead of producing a single metric with value `0`.

### CODE REVIEW

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java [hunk 1]
```diff
@@ -198,6 +202,10 @@ private Context doStart(Context parentContext, REQUEST request, @Nullable Instan
         context = operationListeners[i].onStart(context, attributes, startNanos);
       }
     }
+    // when start and end are not called on the same instrumenter we need to use the operation
+    // listeners that were used during start in end too to correctly handle metrics like
```
Comment (heyams) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java:206]:
```suggestion
    // listeners that were used during start and end to correctly handle metrics like
```

File: testing-common/src/main/java/io/opentelemetry/instrumentation/testing/junit/http/AbstractHttpServerTest.java [hunk 2]
```diff
@@ -353,12 +353,8 @@ void httpServerMetrics() {
               spanData -> assertServerSpan(assertThat(spanData), method, SUCCESS, SUCCESS.status));
         });
 
-    String metricsInstrumentationName = options.metricsInstrumentationName.get();
-    if (metricsInstrumentationName == null) {
-      metricsInstrumentationName = instrumentationName.get();
-    }
     testing.waitAndAssertMetrics(
-        metricsInstrumentationName,
+        instrumentationName.get(),
```
Comment (heyams) [inline testing-common/src/main/java/io/opentelemetry/instrumentation/testing/junit/http/AbstractHttpServerTest.java:357]: 👍

**Review by heyams — APPROVED** [review]

**Review by trask — COMMENTED** [review]
sorry for the late review

do we have any options for fixing this in a way where start/end are always called on the same instrumenter?

I worry about other things not working right when start/end are called on different instances

**Comment by laurit** [comment]
> do we have any options for fixing this in a way where start/end are always called on the same instrumenter?

As far as I understand this only happens when server instrumentation starts the server span and servlet instrumentation ends it when it is an async request. We could avoid it by altering the server instrumentation so that it would not rely on the servlet instrumentation for ending the async spans. Probably undertow instrumentation already works correctly because it ends the server span when the requests completes. In contrast tomcat doesn't because it ends the span when code exits from the `CoyoteAdapter.service` method when async request might still be running, currently it relies on servlet instrumentation to end the span when the async request completes.

**Comment by trask** [comment]
is there a similar issue with `AttributesExtractor.onEnd()` being called on a different set of AttributesExtractors?

**Comment by trask** [comment]
I realized one of my concerns was just that this is "java agent" problems leaking into the instrumenter API, so I hid the behavior behind an internal flag, and the Java agent can opt-in to the behavior for specific instrumenters.

Note: I think I need to also add the context key any time the context key already exists, since otherwise "onEnd" of a nested instrumenter will call the operation listeners of the parent, but going to wait for CI to fail as I'm curious about our test coverage

**Review by trask — APPROVED** [review]

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java [hunk 3]
```diff
@@ -198,6 +204,15 @@ private Context doStart(Context parentContext, REQUEST request, @Nullable Instan
         context = operationListeners[i].onStart(context, attributes, startNanos);
       }
     }
+    if (propagateOperationListenersToOnEnd || context.get(START_OPERATION_LISTENERS) != null) {
+      // when start and end are not called on the same instrumenter we need to use the operation
+      // listeners that were used during start and end to correctly handle metrics like
```
Comment (laurit) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java:209]:
```suggestion
      // listeners that were used during start in end to correctly handle metrics like
```

**Pull request merged**